File: common/src/main/java/dev/aika/bigclever/utils/DNSQuery.java
package dev.aika.bigclever.utils;

import javax.naming.Context;
import javax.naming.directory.Attribute;
import javax.naming.directory.DirContext;
import javax.naming.directory.InitialDirContext;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;

public class DNSQuery {
    public static boolean isSuitableForDNSQuery(String hostname) {
        if ("localhost".equalsIgnoreCase(hostname) || "::1".equals(hostname)) return false;

        try {
            InetAddress address = InetAddress.getByName(hostname);
            return !address.isLoopbackAddress() && !address.isSiteLocalAddress() && !address.isLinkLocalAddress() && !address.isAnyLocalAddress();
        } catch (UnknownHostException ignored) {
        }
        return false;
    }

    public static List<String> queryTXTRecords(String domain) {
        List<String> txtRecords = new ArrayList<>();
        Hashtable<String, String> env = new Hashtable<>();
        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.dns.DnsContextFactory");

        try {
            DirContext ctx = new InitialDirContext(env);
            Attribute attr = ctx.getAttributes(domain, new String[]{"TXT"}).get("TXT");
            if (attr != null) for (int i = 0; i < attr.size(); i++) txtRecords.add((String) attr.get(i));
        } catch (Exception ignored) {
        }
        return txtRecords;
    }
}

File: common/src/main/kotlin/dev/aika/bigclever/utils/DNSQuery.kt
package dev.aika.bigclever.utils

import java.net.InetAddress
import java.net.UnknownHostException
import java.util.*
import javax.naming.Context
import javax.naming.directory.DirContext
import javax.naming.directory.InitialDirContext

object DNSQuery {
    fun isSuitableForDNSQuery(hostname: String): Boolean {
        if ("localhost".equals(hostname, ignoreCase = true) || "::1" == hostname) return false

        try {
            val address = InetAddress.getByName(hostname)
            return !address.isLoopbackAddress && !address.isSiteLocalAddress && !address.isLinkLocalAddress && !address.isAnyLocalAddress
        } catch (ignored: UnknownHostException) {
        }
        return false
    }

    fun queryTXTRecords(domain: String?): List<String> {
        val txtRecords: MutableList<String> = ArrayList()
        val env = Hashtable<String, String>()
        env[Context.INITIAL_CONTEXT_FACTORY] = "com.sun.jndi.dns.DnsContextFactory"

        try {
            val ctx: DirContext = InitialDirContext(env)
            val attr = ctx.getAttributes(domain, arrayOf("TXT"))["TXT"]
            if (attr != null) for (i in 0 until attr.size()) txtRecords.add(attr[i] as String)
        } catch (ignored: Exception) {
        }
        return txtRecords
    }
}